src/lib/crypto/snarky_tests/snarky_tests.ml

```diff
@@ -0,0 +1,274 @@
+
+    Impl.Field.Assert.equal res two ;
+    Impl.Field.Assert.not_equal res Impl.Field.one ;
+    Impl.Field.Assert.non_zero res
+end
+
+module RangeCircuits = struct
+  let input_typ = Impl.Field.typ
+
+  let return_typ = Impl.Typ.unit
+
+  let lte (x : Impl.Field.t) () =
+    Impl.Field.Assert.lte ~bit_length:2 x (Impl.Field.of_int 3) ;
+    ()
+
+  let gte (x : Impl.Field.t) () =
+    Impl.Field.Assert.gte ~bit_length:2 x (Impl.Field.of_int 3) ;
+    ()
+
+  let gt (x : Impl.Field.t) () =
+    Impl.Field.Assert.gt ~bit_length:2 x (Impl.Field.of_int 3) ;
+    ()
+
+  let lt (x : Impl.Field.t) () =
+    Impl.Field.Assert.lt ~bit_length:2 x (Impl.Field.of_int 3) ;
+    ()
+end
+
+module TernaryCircuit = struct
+  let input_typ = Impl.Boolean.typ
+
+  let return_typ = Impl.Typ.unit
+
+  let main (x : Impl.Boolean.var) () =
+    let two = Impl.Field.of_int 2 in
+    let res = Impl.if_ x ~typ:Impl.Field.typ ~then_:two ~else_:Impl.Field.one in
+    Impl.Field.Assert.equal res two
+end
+
+module PublicOutput = struct
+  let input_typ = Impl.Boolean.typ
+
+  let return_typ = Impl.Boolean.typ
+
+  let main (x : Impl.Boolean.var) () : Impl.Boolean.var =
+    Impl.Boolean.(x && Impl.Boolean.true_)
+end
+
+let circuit_tests =
+  [ ( "boolean circuit"
+    , `Quick
+    , check_json ~input_typ:BooleanCircuit.input_typ
+        ~return_typ:BooleanCircuit.return_typ ~circuit:BooleanCircuit.main
+        "simple.json" )
+  ; ( "circuit with field arithmetic"
+    , `Quick
+    , check_json ~input_typ:FieldCircuit.input_typ
+        ~return_typ:FieldCircuit.return_typ ~circuit:FieldCircuit.main
+        "field.json" )
+  ; ( "circuit with range check (less than equal)"
+    , `Quick
+    , check_json ~input_typ:RangeCircuits.input_typ
+        ~return_typ:RangeCircuits.return_typ ~circuit:RangeCircuits.lte
+        "range_lte.json" )
+  ; ( "circuit with range check (greater than equal)"
+    , `Quick
+    , check_json ~input_typ:RangeCircuits.input_typ
+        ~return_typ:RangeCircuits.return_typ ~circuit:RangeCircuits.gte
+        "range_gte.json" )
+  ; ( "circuit with range check (less than)"
+    , `Quick
+    , check_json ~input_typ:RangeCircuits.input_typ
+        ~return_typ:RangeCircuits.return_typ ~circuit:RangeCircuits.lt
+        "range_lt.json" )
+  ; ( "circuit with range check (greater than)"
+    , `Quick
+    , check_json ~input_typ:RangeCircuits.input_typ
+        ~return_typ:RangeCircuits.return_typ ~circuit:RangeCircuits.gt
+        "range_gt.json" )
+  ; ( "circuit with ternary operator"
+    , `Quick
+    , check_json ~input_typ:TernaryCircuit.input_typ
+        ~return_typ:TernaryCircuit.return_typ ~circuit:TernaryCircuit.main
+        "ternary.json" )
+  ; ( "circuit with public output"
+    , `Quick
+    , check_json ~input_typ:PublicOutput.input_typ
+        ~return_typ:PublicOutput.return_typ ~circuit:PublicOutput.main
+        "output.json" )
+  ]
+
+(* API tests *)
+
+let get_hash_of_circuit () =
```

Comment:
This seems to be repeated

Reply:
one is using the monadic API and the other uses the imperative API